File: core/metrics.py
"""
Metrics calculation for OCR evaluation.
Computes precision, recall, CRR (Character Recognition Rate), and F1 score.
"""

from .utils import calculate_crr


def calculate_metrics(matches):
    """
    Calculate OCR evaluation metrics from word matches.

    Metrics:
    - Precision: exact_matches / total_ocr_words (only exact matches count)
    - Recall: exact_matches / total_gt_words (only exact matches count)
    - Average CRR: Character Recognition Rate across all matched pairs (exact + fuzzy)
    - F1 Score: Harmonic mean of precision and recall

    Args:
        matches (list): List of match tuples from matcher.match_words()
                       (gt_word, ocr_word, edit_distance, match_type)

    Returns:
        dict: Dictionary with keys:
            - precision (float): 0.0 to 1.0
            - recall (float): 0.0 to 1.0
            - avg_crr (float): 0.0 to 1.0
            - f1_score (float): 0.0 to 1.0
            - exact_matches (int): Count of exact matches
            - fuzzy_matches (int): Count of fuzzy matches
            - total_gt_words (int): Total words in ground truth
            - total_ocr_words (int): Total words in OCR output
            - unmatched_gt (int): GT words with no match
            - unmatched_ocr (int): OCR words with no match
    """
    # Separate matches by type
    exact_matches = [m for m in matches if m[3] == 'exact']
    fuzzy_matches = [m for m in matches if m[3] == 'fuzzy']
    gt_only_matches = [m for m in matches if m[3] == 'gt_only']
    ocr_only_matches = [m for m in matches if m[3] == 'ocr_only']

    # Count totals
    exact_count = len(exact_matches)
    fuzzy_count = len(fuzzy_matches)
    total_gt_words = exact_count + fuzzy_count + len(gt_only_matches)
    total_ocr_words = exact_count + fuzzy_count + len(ocr_only_matches)

    # Calculate Precision and Recall (only exact matches count)
    # This aligns with Word Error Rate (WER) standards
    precision = exact_count / total_ocr_words if total_ocr_words > 0 else 0.0
    recall = exact_count / total_gt_words if total_gt_words > 0 else 0.0

    # Calculate F1 Score
    if precision + recall > 0:
        f1_score = 2 * (precision * recall) / (precision + recall)
    else:
        f1_score = 0.0

    # Calculate Average CRR for all matched pairs (exact + fuzzy)
    # This shows character-level accuracy for recognized words
    crr_scores = []

    # Exact matches have CRR = 1.0 (edit distance = 0)
    for gt_word, ocr_word, edit_dist, match_type in exact_matches:
        crr = calculate_crr(gt_word, ocr_word)
        crr_scores.append(crr)

    # Fuzzy matches have CRR based on edit distance
    for gt_word, ocr_word, edit_dist, match_type in fuzzy_matches:
        crr = calculate_crr(gt_word, ocr_word)
        crr_scores.append(crr)

    avg_crr = sum(crr_scores) / len(crr_scores) if crr_scores else 0.0

    return {
        'precision': precision,
        'recall': recall,
        'avg_crr': avg_crr,
        'f1_score': f1_score,
        'exact_matches': exact_count,
        'fuzzy_matches': fuzzy_count,
        'total_gt_words': total_gt_words,
        'total_ocr_words': total_ocr_words,
        'unmatched_gt': len(gt_only_matches),
        'unmatched_ocr': len(ocr_only_matches)
    }


def format_metrics_for_display(metrics):
    """
    Format metrics as percentages for display.

    Args:
        metrics (dict): Metrics dictionary from calculate_metrics()

    Returns:
        dict: Formatted metrics with percentage strings
    """
    return {
        'precision': f"{metrics['precision'] * 100:.2f}%",
        'recall': f"{metrics['recall'] * 100:.2f}%",
        'avg_crr': f"{metrics['avg_crr'] * 100:.2f}%",
        'f1_score': f"{metrics['f1_score'] * 100:.2f}%",
        'exact_matches': metrics['exact_matches'],
        'fuzzy_matches': metrics['fuzzy_matches'],
        'total_gt_words': metrics['total_gt_words'],
        'total_ocr_words': metrics['total_ocr_words'],
        'unmatched_gt': metrics['unmatched_gt'],
        'unmatched_ocr': metrics['unmatched_ocr']
    }

File: static/js/metrics.js
/**
 * Metrics calculation for OCR evaluation.
 * Computes precision, recall, CRR (Character Recognition Rate), and F1 score.
 */

/**
 * Calculate OCR evaluation metrics from word matches.
 *
 * Metrics:
 * - Precision: exact_matches / total_ocr_words (only exact matches count)
 * - Recall: exact_matches / total_gt_words (only exact matches count)
 * - CRR: Character Recognition Rate = 1 - CER (document-level, includes all characters)
 * - F1 Score: Harmonic mean of precision and recall
 *
 * @param {Array} matches - List of match tuples from matchWords()
 * @returns {Object} Dictionary with metrics
 */
function calculateMetrics(matches) {
    // Separate matches by type
    const exactMatches = matches.filter(m => m[3] === 'exact');
    const gtOnlyMatches = matches.filter(m => m[3] === 'gt_only');
    const ocrOnlyMatches = matches.filter(m => m[3] === 'ocr_only');

    // Count totals
    const exactCount = exactMatches.length;
    const totalGtWords = exactCount + gtOnlyMatches.length;
    const totalOcrWords = exactCount + ocrOnlyMatches.length;

    // Calculate Precision and Recall (only exact matches count)
    const precision = totalOcrWords > 0 ? exactCount / totalOcrWords : 0.0;
    const recall = totalGtWords > 0 ? exactCount / totalGtWords : 0.0;

    // Calculate F1 Score
    let f1Score = 0.0;
    if (precision + recall > 0) {
        f1Score = 2 * (precision * recall) / (precision + recall);
    }

    // Calculate document-level CRR (Character Recognition Rate)
    let totalCharErrors = 0;
    let totalGtChars = 0;

    // For exact matches: 0 errors (words are identical)
    for (const [gtWord] of exactMatches) {
        // No errors for exact matches
        totalGtChars += gtWord.length;
    }

    // For unmatched GT words: all characters are errors (deletions - OCR missed them)
    for (const [gtWord] of gtOnlyMatches) {
        totalCharErrors += gtWord.length;
        totalGtChars += gtWord.length;
    }

    // For unmatched OCR words: all characters are errors (insertions - OCR hallucinated them)
    for (const [, ocrWord] of ocrOnlyMatches) {
        totalCharErrors += ocrWord.length;
    }

    // Calculate CRR (Character Recognition Rate = 1 - CER)
    // CRR can be negative if OCR produces more errors than GT characters
    let avgCrr = 0.0;
    if (totalGtChars > 0) {
        const cer = totalCharErrors / totalGtChars;
        avgCrr = 1.0 - cer;
    }

    return {
        precision: precision,
        recall: recall,
        avg_crr: avgCrr,
        f1_score: f1Score,
        exact_matches: exactCount,
        total_gt_words: totalGtWords,
        total_ocr_words: totalOcrWords,
        unmatched_gt: gtOnlyMatches.length,
        unmatched_ocr: ocrOnlyMatches.length,
        // Additional data for tooltips
        char_errors: totalCharErrors,
        total_gt_chars: totalGtChars
    };
}

/**
 * Format metrics as percentages for display.
 *
 * @param {Object} metrics - Metrics object from calculateMetrics()
 * @returns {Object} Formatted metrics with percentage strings
 */
function formatMetricsForDisplay(metrics) {
    return {
        precision: `${(metrics.precision * 100).toFixed(2)}%`,
        recall: `${(metrics.recall * 100).toFixed(2)}%`,
        avg_crr: `${(metrics.avg_crr * 100).toFixed(2)}%`,
        f1_score: `${(metrics.f1_score * 100).toFixed(2)}%`,
        exact_matches: metrics.exact_matches,
        total_gt_words: metrics.total_gt_words,
        total_ocr_words: metrics.total_ocr_words,
        unmatched_gt: metrics.unmatched_gt,
        unmatched_ocr: metrics.unmatched_ocr
    };
}

/**
 * Metrics utility functions for tooltips and display.
 * Attached to window for global access.
 */
window.MetricsUtils = {
    /**
     * Generate tooltip text showing how a metric was calculated.
     *
     * @param {string} metricName - Name of the metric ('precision', 'recall', 'f1_score', 'avg_crr')
     * @param {Object} metrics - Metrics object from calculateMetrics()
     * @returns {string} Tooltip text with formula and calculation
     */
    getTooltip(metricName, metrics) {
        const exact = metrics.exact_matches;
        const totalGt = metrics.total_gt_words;
        const totalOcr = metrics.total_ocr_words;
        const charErrors = metrics.char_errors;
        const totalGtChars = metrics.total_gt_chars;

        switch (metricName) {
            case 'precision':
                return `Precision = Exact Matches / Total OCR Words\n= ${exact} / ${totalOcr}\n= ${(metrics.precision * 100).toFixed(2)}%`;

            case 'recall':
                return `Recall = Exact Matches / Total GT Words\n= ${exact} / ${totalGt}\n= ${(metrics.recall * 100).toFixed(2)}%`;

            case 'f1_score':
                const p = (metrics.precision * 100).toFixed(2);
                const r = (metrics.recall * 100).toFixed(2);
                return `F1 Score = 2 × (Precision × Recall) / (Precision + Recall)\n= 2 × (${p}% × ${r}%) / (${p}% + ${r}%)\n= ${(metrics.f1_score * 100).toFixed(2)}%`;

            case 'avg_crr':
                const cer = totalGtChars > 0 ? (charErrors / totalGtChars * 100).toFixed(2) : '0.00';
                return `CRR = 1 - (Character Errors / Total GT Chars)\n= 1 - (${charErrors} / ${totalGtChars})\n= 1 - ${cer}%\n= ${(metrics.avg_crr * 100).toFixed(2)}%`;

            default:
                return '';
        }
    },

    /**
     * Apply tooltip to an existing DOM element.
     *
     * @param {HTMLElement} element - The DOM element to add tooltip to
     * @param {string} metricName - Name of the metric
     * @param {Object} metrics - Metrics object from calculateMetrics()
     */
    applyTooltip(element, metricName, metrics) {
        element.classList.add('metric-with-tooltip');

        // Create tooltip element
        const tooltip = document.createElement('div');
        tooltip.className = 'metric-tooltip';
        tooltip.textContent = this.getTooltip(metricName, metrics);
        element.appendChild(tooltip);
    },

    /**
     * Create a table cell with metric value and tooltip.
     *
     * @param {number} value - The metric value (0.0-1.0)
     * @param {string} metricName - Name of the metric
     * @param {Object} metrics - Metrics object from calculateMetrics()
     * @returns {HTMLTableCellElement} The created td element
     */
    createMetricCell(value, metricName, metrics) {
        const td = document.createElement('td');
        td.textContent = Utils.formatPercentage(value);
        this.applyTooltip(td, metricName, metrics);
        return td;
    }
};
